packages/tangle/parser.go

```diff
@@ -49,6 +51,7 @@ func NewParser() (result *Parser) {
 
 	// add builtin filters
 	result.AddBytesFilter(NewRecentlySeenBytesFilter())
+	result.AddMessageFilter(NewMessageSyntacticalValidator())
```

Comment:
Why introducing a filter to do this kind of job? Syntactical checks on the previous message layout were performed without.

Reply:
The reasons are:
1. Currently other validation checks are done via filters. A maintainer would expect all validations to be done this way.
2. Were you referring to the validation done while parsing the message (were you?) imo adding checks there would just conflate between parsing and validation.
3. The filter framework currently creates events for rejection that we can hook into
4. If for some reason an api would be added to send a message to node directly (not just payload) this would become handy.

Reply:
Checks before were performed without the use of a filter; so the code is now checking twice: with the "old" mechanism and this new filter. I am not sure that the filter is the best approach, because the checks must be performed on parsing anyway, so these are essentially duplicated checks.

Reply:
I understand syntactical validation to be part of the parsing to make sure that the bytes you're reading actually correspond to the bytes you're expecting to read (ie you know that what you're reading is a correct message). This includes loading an object from disk as well as receiving it from a neighbor and API. As such, this needs to happen somewhere within a call of `MessageFromBytes` to cover all these cases. A filter then simply does these checks redundantly, and is thus not necessary anymore.

Reply:
I just solved a real silly bug that drove me crazy yesterday. It was really an obvious one in retrospect but it did cost me an entire day of work, There is a good chance that if I would've continued to mismash parsing and validation together this bug would've been left undiscovered...

The bug was that the parsed `parentsBlocksCount` was never assigned back to the `message` struct.
It was a bit elusive to me because since I saw it correctly parsed I assumed it was assigned to the message,

Now if we add more validation logic to the parsing code we will make the function longer, thus harder to maintain.
If I would do it the way you guys instructed me now then I might have added a call at the end of `MessageFromBytes` called `validate` to account for all the less trivial rules (uniqueness, lexicographical order), so I won't have reptitions. This approach is likely to lead to more bugs! This would be because `validate` would be missing validation logic that is implicit in `parseMessageBytes`.

Since `validate` is in package scope and we like to have flat design where everything is accessible it may get reused in the future. *Currently* you may say there is no need for it to get reused so it doesn't matter. But is it *guaranteed* to  stay this way? One should always assume he won't know the code requirements in the future.

So I believe that parsing and validation are actually two seperate things and we should resist the temptation of unifying them. As long as we adhere to certain principles (keeping functions short, with 1 clear-cut responsibility, and reusable) we will probably end up with a more maintainable and less bug-prone code...

Sorry for such a long writeup, but I think this kind of discussion saves time in the end :-)